### PR TITLE
Add instrgen implementation

### PR DESCRIPTION
**Pull request by pdelewski**

### CODE REVIEW

File: instrgen/driver/main.go
Comment (dineshg13) [inline instrgen/driver/main.go:50]: Is this useful for user ?

File: instrgen/driver/main.go
Comment (dineshg13) [inline instrgen/driver/main.go:104]: Can we use `switch` in case of `if`

File: instrgen/lib/callgraph.go
Comment (dineshg13) [inline instrgen/lib/callgraph.go:56]: Can u add more meaningful println , like `finding packages` etc ?

**Review by dineshg13 — COMMENTED** [review]
Can you please add how to run this instrument ? And please trim printlns. 

When i run in macos or linux am getting this error ( go version go1.19.5 darwin/arm64 ) 

```
./instrgen --inject ./testdata/basic/fib.go ./                                                                                                                                                                                                                                                                                                                                                                                                                                   ─╯
autotel compiler
FindInterfaces
FindFuncDecls
BuildCallGraph

        child parent

2023/02/05 07:29:29 err: fork/exec /opt/homebrew/bin/go: not a directory: stderr:
```

**Comment by pdelewski** [comment]
> Can you please add how to run this instrument ? And please trim printlns.
> 
> When i run in macos or linux am getting this error ( go version go1.19.5 darwin/arm64 )
> 
> ```
>` ./instrgen --inject ./testdata/basic/fib.go ./  `                                                                                                                                                                                                                                                                                                                                                                                                                                 ─╯
> autotel compiler
> FindInterfaces
> FindFuncDecls
> BuildCallGraph
> 
>         child parent
> 
> 2023/02/05 07:29:29 err: fork/exec /opt/homebrew/bin/go: not a directory: stderr:
> ```

@dineshg13 works on packages so second parameter must be path not file, so in this case ` ./instrgen --inject ./testdata/basic/ ./  ` will work, however it will inject into all files in this directory. I grouped few simple programs under one sub-directory due to test execution speedup

File: instrgen/lib/callgraph.go
Comment (pdelewski) [reply instrgen/lib/callgraph.go:56]: Yes

File: instrgen/driver/main.go
Comment (pdelewski) [reply instrgen/driver/main.go:104]: Yes

File: instrgen/driver/main.go
Comment (pdelewski) [reply instrgen/driver/main.go:50]: This is useful for debugging purposes. I left just for now, but of course we can get rid of this printfs.

File: instrgen/rtlib/rtlib.go
Comment (pdelewski) [reply instrgen/rtlib/rtlib.go:82]: Thx. That's kind of leftover. I have to think what to do with it.

**Comment by MrAlias** [comment]
SIG meeting: plan is to add a readme to this.

**Review by MrAlias — APPROVED** [review]
This looks good enough to start with. I've found a fair amount of issues, but I don't think any of them fundamentally block this from being merged as a starting point.

Thanks for getting this all together @pdelewski! :tada:

File: CHANGELOG.md
Comment (pdelewski) [reply CHANGELOG.md]: done

**Review by dineshg13 — APPROVED** [review]
This is a good start. I was able to instrument a sample HTTP server using this. @pdelewski Thanks for the contribution.

**Comment by pdelewski** [comment]
@MrAlias Seems that we can merge it.

**Comment by MrAlias** [comment]
Merging as this has 2 reviews from the @open-telemetry/go-instrumentation-approvers group.

**Pull request merged**
